oauth2.md

```diff
@@ -2,65 +2,164 @@
 
 # Oauth2 Features
 
-Stormpath provides two Oauth2 workflows:
+Stormpath provides two Oauth2 workflows, client credential and password.  In
+this document we discuss how these should be exposed in the developer's web
+application.
 
-### Client Credentials
 
-In this workflow, an api key and secret is provisioned for an account.  These
-credentials can be exchanged for an access token by making a POST request
-to `/oauth/token`.  The request must contain the header
-`Authorization: Basic <base64UrlEncoded(apiKeyId:apiKeySecret)` and the body
-must contain `grant_type=client_credentials`
 
-The underlyhing Stormpath SDK is responsible for generating the access token
-and verifying that the API key & secret is valid, and that the account is not
-disabled
+### OAuth2 Configuration Options
 
-The develoepr can specify these options for th client credentials workflow:
+The OAuth2 endpoint can be configured, or disabled entirely:
 
 ```yaml
 web:
   oauth2:
     enabled: true
     uri: "/oauth/token"
+```
+
+#### web.oauth2.enabled
+
+By default we accept POSTS to this token uri and respond according to the
+grant type.  If the developer sets this value to false we should not attach any
+handler to this uri, allowing the framework to return it's default 404 response.
+
+If enabled and the grant type is not `client_credentials` or `password`, then we
+should return the OAuth-compliant error code:
+
+    HTTP/1.1 400 Bad Request
+    Content-Type: application/json;charset=UTF-8
+    Cache-Control: no-store
+    Pragma: no-cache
+
+    {
+      "error": "unsupported_grant_type"
+    }
+
+If no grant type is specified, the error should be `invalid_request`
+
+## Client Credentials Grant Flow
+
+The product guide for this feature can be found here:
+
+https://docs.stormpath.com/guides/api-key-management/
+
+In this workflow, an api key and secret is provisioned for a stormpath account.
+These credentials can be exchanged for an access token by making a POST request
+to `/oauth/token` on the web application.  The request must look like this:
+
+````
+POST /oauth/token
+Authorization: Basic <base64UrlEncoded(apiKeyId:apiKeySecret)>
+
+grant_type=client_credentials
+````
+
+For this flow the underlying Stormpath SDK is used for generating the access
+token and verifying that the API key & secret is valid, and that the account is
+not disabled.  The response is an access token response, but this flow does not
+return a refresh token (only an access token):
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json;charset=UTF-8
+Cache-Control: no-store
+Pragma: no-cache
+
+{
+  "access_token":"2YotnFZFEjr1zCsicMWpAA...",
+  "expires_in":3600,
+  "token_type":"Bearer"
+}
+```
+
+### Client Credentials Options
+
+```yaml
+web:
+  oauth2:
     client_credentials:
```

Comment:
Thinking out loud here, but not entirely sure...

Instead of `client_credentials:  enabled: true`, what about:

```yaml
web:
    oauth2:
        grantTypes:
            - client_credentials:
                  enabled: true
            - password
```

i.e. a list?  or should it still be name-value pairs?  Thoughts?

Also, should each grant type have its own ttl setting?  Or would that be shared:

```yaml
web:
    oauth2:
        ttl: 3600
        grantTypes:
            - client_credentials:
                  enabled: true
            - password
```

?